pytest.ini

```diff
@@ -12,6 +12,90 @@ markers =
+  # * `awx/main/tests/unit/utils/test_common.py::TestHostnameRegexValidator::test_bad_call_with_inverse`
+  once:assertion is always true, perhaps remove parentheses\?:pytest.PytestAssertRewriteWarning:
+
+  # FIXME: Figure this out, fix and then delete the entry. It's not entirely
+  # FIXME: clear what emits it and where.
+  once:Pagination may yield inconsistent results with an unordered object_list. .class 'awx.main.models.workflow.WorkflowJobTemplateNode'. QuerySet.:django.core.paginator.UnorderedObjectListWarning:django.core.paginator
+
+  # FIXME: Figure this out, fix and then delete the entry.
+  once::django.core.paginator.UnorderedObjectListWarning:rest_framework.pagination
+
+  # FIXME: Replace use of `distro.linux_distribution()` via a context manager
+  # FIXME: in `awx/main/analytics/collectors.py` and then delete the entry.
+  once:distro.linux_distribution.. is deprecated. It should only be used as a compatibility shim with Python's platform.linux_distribution... Please use distro.id.., distro.version.. and distro.name.. instead.:DeprecationWarning:awx.main.analytics.collectors
+
+  # FIXME: Figure this out, fix and then delete the entry.
+  once:\nUsing ProtocolTypeRouter without an explicit "http" key is deprecated.\nGiven that you have not passed the "http" you likely should use Django's\nget_asgi_application...:DeprecationWarning:awx.main.routing
+
+  # FIXME: Figure this out, fix and then delete the entry.
+  once:Channel's inbuilt http protocol AsgiHandler is deprecated. Use Django's get_asgi_application.. instead.:DeprecationWarning:channels.routing
+
+  # FIXME: Use `codecs.open()` via a context manager
+  # FIXME: in `awx/main/utils/ansible.py` to close hanging file descriptors
+  # FIXME: and then delete the entry.
+  once:unclosed file <_io.BufferedReader name='[^']+'>:ResourceWarning:awx.main.utils.ansible
+
+  # FIXME: Use `open()` via a context manager
+  # FIXME: in `awx/main/tests/unit/test_tasks.py` to close hanging file
+  # FIXME: descriptors and then delete the entry.
+  once:unclosed file <_io.TextIOWrapper name='[^']+' mode='r' encoding='UTF-8'>:ResourceWarning:awx.main.tests.unit.test_tasks
+
+  # FIXME: Use `open()` via a context manager
+  # FIXME: in `awx/main/tests/unit/test_tasks.py` to close hanging file
+  # FIXME: descriptors and then delete the entry.
+  once:unclosed file <_io.BufferedReader name='[^']+'>:ResourceWarning:awx.main.tests.unit.test_tasks
+
+  # FIXME: Use `open()` via a context manager
+  # FIXME: in `awx/main/tests/unit/test_tasks.py` to close hanging file
+  # FIXME: descriptors and then delete the entry.
+  once:unclosed file <_io.TextIOWrapper name='[^']+' mode='r' encoding='UTF-8'>:ResourceWarning:_pytest.python
+
+  # FIXME: Use `open()` via a context manager
+  # FIXME: in `awx/main/utils/common.py` to close hanging file descriptors
+  # FIXME: and then delete the entry.
+  once:unclosed file <_io.BufferedWriter name='[^']+'>:ResourceWarning:awx.main.utils.common
```

Comment:
This feels like re-arranging deck chairs on the Titanic. Great that your change gets us exposed to new warnings, but it will make dependency upgrades take longer, which already happen too infrequently, and it almost guarantees that none of these issues will be worked. And they really should be.

The methods here are fine, but I'd like to at least have at least a plan to knock out the borderline absurd ones in the list.

Reply:
I skipped addressing them because that wasn't my goal. I wanted to show how to configure pytest properly. Fixing them can be planned for whenever tech debt is worked on.
I don't think that it causes more work, it just shifts the time when mandatory changes need to be made. Besides, you can always add more ignores if you want to postpone.

Reply:
Is this PR, itself, not tech debt? And like I said, I'm fine to file followup, but "whenever tech debt is worked on" doesn't happen. Even a catch-all issue referencing your list will never get attention.

I meant it to comment on the warning about hanging file descriptors because absolutely nobody will disagree with that change, for example:

https://github.com/ansible/awx/blob/69baa739fae9a0523b1c1853449614ba8256d294/awx/asgi.py#L27-L28

I don't want to just have this added to a forever-ignore list, unless, for some reason I'm wrong about the complexity of it. Again, filing issues about the specific issues being ignored might be a workable approach.

Reply:
This doesn't just document existing issues, it changes the way of working by introducing an ignore-list. The reason the warnings were not silenced is because of a shared belief that they are important and should have code changes to resolve. This disposition doesn't necessarily change... but if you're not clear, people will do the wrong thing. I don't trust that without an example, so I created one:

https://github.com/webknjaz/ansible--awx/pull/2

Getting a new process for contributions, without getting some contributions (separate PRs is fine) is not something I'm really happy with. But here, I'm seeking an explicit consensus between you and me, and specific followup issues.

Reply:
Okay, that makes sense in a way. Though, I don't think it changes the process. Previously, deprecation warnings would have to be addressed *at some point* and they were invisible (by default `UserWarning` instances are printed on the terminal, while many things, `DeprecationWarning` and `FutureDeprecationWarning` included never show up anywhere). To reveal those, one has to know upfront that they have to invoke everything using `python -Werror -Im thing` (or similar via an env var). And then, those places would have to be fixed. The problem is that with invisible warnings that accumulate over time, one would have to be aware of all these. And the process was fixing the errors too late when upstreams would replace warnings with hard failures.
The current setup makes these things visible ahead of time which is a pretty much standard best practice in the external Python world. So to me saying that the process changed would be too strong. I'd rather say that it became more transparent.

Reply:
Remaining easily workable categories I see are: 

 - FIXME: Delete this entry once `USE_L10N` use is removed
 - remove distutils from repo (only use, not dep)
 - Delete this entry once naive dates aren't passed to DB lookup
 - 'index_together' is deprecated. Use 'Meta.indexes'
 - Using QuerySet.iterator.. after prefetch_related.. without specifying chunk_size
 - remove always-true assertions
 - fix Pagination may yield inconsistent results with an unordered object_list. .class
 - Channel's inbuilt http protocol AsgiHandler is deprecated. Use Django's get_asgi_application
 - Use `codecs.open()` via a context manager

I would like these to be filed, so @djyasin can have the available for the next sprint. I would have liked to have them even for story pointing, but it might be too late for the current round.

Reply:
@AlanCoding fair. I was also postponing fixing things like `open()` under CM, since those are straightforward and could be a good learning opportunity for folks to try addressing them as opposed to more involved problems.

Reply:
By the way, in the linked place in asgi, a more elegant solution would be `pathlib.Path('/var/lib/awx/.tower_version').read_text().strip()` even. This is to say that every place needs to be inspected additionally, beyond the generic suggestion I've put in the comments.